File: leetcode/kotlin/50. Pow(x, n).kt
internal class Main {
    fun myPow(x: Double, n: Int): Double {
        if (n == 0) return 1.0
        val half = myPow(x, n / 2)
        return if (n % 2 == 0) half * half else if (n > 0) half * half * x else half * half / x
    } // a^m + a^n = a^(m+n)
    // a^-m + a^-n = a^-(m+n)
}

File: leetcode/python/50. Pow(x, n).py
def myPow(x,n):
  if n==0:
    return 1.0
  half = myPow(x,n//2)
  if n%2==0:
    return half * half
  elif n>0:
    return half * half * x
  else:
    return half * half/x

print(myPow(2.00000,10))